### ISSUE TITLE
WIP: Improved TypeScript definitions by generating from Flow source.

### ISSUE THREAD

**Issue by alloy** [react-native-community/discussions-and-proposals#201]
🚧 This is a work-in-progress and will be updated as I go. No need for input just yet.

Re https://github.com/react-native-community/discussions-and-proposals/projects/1#card-31965228

### Introduction

Improved TypeScript definitions for React Native by converting Flow types automatically.

### The Core of It

Thus far, [the TS typings for RN](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-native/index.d.ts) have been maintained by hand by third-party community members, in a ‘as needed by maintainers basis’. This means that the typings are generally out-of-date, either removed APIs are still typed or new APIs have not been added yet, or they are plain wrong.

In order to provide a better solution for both MS’ internal needs as well as external TS users; the question posed is, to what extent can we automatically convert the Flow typings to TS? (Some manual massaging of conversion process may be required and is ok, as these could be applied in the form of patches we maintain.)

The Flow team is interested in helping out, where needed. They would like to see an example of the Flow output we’d need; e.g. what comments are preserved, how do we deal with type aliases, imported types (if we are preserving aliases), etc. However, before going down this road, I want to see what prior art exists and if any of these tools is already up for the task (or at least mature enough).

### Discussion points

Notes of trying out these tools will be kept [here](https://github.com/alloy/react-native/blob/flow-to-ts-2/typescript/Notes.md).

* [Khan Academy’s flow-to-ts](https://github.com/Khan/flow-to-ts): Unclear what the original purpose was of the tool, but from the README seems to position itself as wanting to be a maintained tool. It also does some transformations specific to React’s TS typings. At least [one person reacted](https://twitter.com/kenneth_skovhus/status/1224714558016774147) saying that they tested all other offerings and this one being the best. Last work was done the end of 2019.

  First simple test case done with this tool, e.g.: [Flow](https://github.com/alloy/react-native/blob/053511535ea209483332c52f9f92abbd17f5cf50/Libraries/ActionSheetIOS/ActionSheetIOS.js) vs [TS](https://github.com/alloy/react-native/blob/0850e5bbbe5fb23b0a28143de125a4f78cf04f36/typescript/flow-to-ts/Libraries/ActionSheetIOS/ActionSheetIOS.d.ts).

* [@bcherny & @benjie’s flow2typescript](https://github.com/benjie/flow2typescript): Was used by @benjie to [convert a big project](https://twitter.com/Benjie/status/1225086334617911296) _once_. Last work was done the end of 2019.

  Not tested yet.

* [Quizlet’s babel-plugin-flow-to-typescript](https://github.com/Kiikurage/babel-plugin-flow-to-typescript): From their [blog post](https://medium.com/tech-quizlet/now-or-never-migrating-300k-loc-from-flow-to-typescript-at-quizlet-d3bae5830a1) it looks like it was meant for a one time conversion. Last work was done the end of 2019.

  Not tested yet.

* [@Naturalclar’s f2t ](https://github.com/nd-02110114/f2t): [Sounds like](https://twitter.com/natural_clar/status/1224702533161115649) it was a one-off conversion tool and seems more naive than the other solutions.

  Not tested yet.

**Comment by alloy** [comment]
`flow-to-ts` was the most full-featured for our needs, so I continued that prototyping in a repo of its own: https://github.com/alloy/rn2dts.

Next step is to see how fast I can get rid of the majority of the `tsc` errors.

**Comment by alloy** [comment]
In an effort to try to get TypeScript definitions for React Native consistent with the cononical Flow types, [I have been looking into](https://github.com/alloy/rn2dts) the possibility of automating this process.

My initial approach was to see how far I could get _without_ any help needed from Facebook’s side.

<details>
<summary>The conversion process looks something like this:</summary>

1. Transform Flow codebase to TypeScript by leveraging [existing tooling](https://github.com/Khan/flow-to-ts) with a few [additional changes](https://github.com/alloy/rn2dts/blob/master/patches/%40khanacademy%2Bflow-to-ts%2B0.1.5.patch), which mostly replaces Flow utility types with TypeScript ones. (This tooling is presumably meant for people migrating from Flow to TypeScript and thus it is ok for these tools to not deliver perfect conversions.)
1. Replace `$ObjMap` usage with a mapped type, as RN’s usage of `$ObjMap` is trivial (return types are static).
1. Replace `import type {Foo} from "./some-module"` with `import {Foo} from "./some-module"`.
1. Replace `module.exports = {…}` with `export default {…}`.
1. Replace `const Foo = require("./some-module")` calls with `import Foo from "./some-module"` statements and hoist+dedupe them.
1. Replace `import typeof Foo from "./some-module"` with `import _Foo from "./some-module"; type Foo = typeof _Foo;`.
1. Remove bodies from functions that have explicit return typing.
1. Remove class members whose name starts with an underscore.
1. Remove `React.Component` life-cycle related class members from subclasses and other implementation details such as its `state` typing.
1. And finally feed this to `tsc` to compile, infer untyped return types, and emit declarations.
</details>

Much of this actually works surprisingly well, albeit slightly convoluted, but I did hit a few snags related to CommonJS imports/exports:

<details>
<summary>default export of single item and <code>import typeof</code> (fixable in TS type system)</summary>

For example, a CommonJS export like the following:

```js
const Foo = {…}
module.exports = Foo
```

…converted to:

```js
const Foo = {…}
default export Foo
```

👍 …will work fine with TypeScript as a value:

```ts
import Foo from "./some-module"
// Foo is a value
console.log(Foo)
```

👍  …and when converted from `import typeof`:

```ts
import _Foo from "./some-module"
// The type of the object shape
type Foo = typeof _Foo
```

----

Equally a single exported class _as a **value**_ will work fine:

```js
class Foo {…}
module.exports = Foo
```

…converted to:

```js
class Foo {…}
default export Foo
```

👍 …will work fine with TypeScript as a value:

```ts
import Foo from "./some-module"

// Foo is a value
console.log(new Foo())
```

👎 …however, in the case where `import typeof Foo from "./some-module"` was converted, we end up with the type of the _static_ class members. Luckily this is fixable in a static manner for both objects and classes, albeit in a roundabout way:

```ts
type ClassInstanceOrType<T> = T extends { new(...args: any): any } ? InstanceType<T> : T

import _Foo from "./some-module"
// Refers to the instance type of class Foo
type Foo = ClassInstanceOrType<typeof _Foo>
```

</details>

<details>
<summary>export of multiple items and importing specific items (not easily fixable)</summary>

For example, a CommonJS export like the following:

```js
const Foo = {…}

class Bar {…}

module.exports = { Foo, Bar }
```

…when converted to a single default export:

```ts
const Foo = {…}

class Bar {…}

export default { Foo, Bar }
```

…means that TypeScripts users won’t be able to do:

```ts
import { Foo } from "./some-module"
```

…instead they would have to do:

```ts
import SomeModule from "./some-module"
const { Foo } = SomeModule
```

This is, in my opinion, too counterintuitive to be acceptable.

Ideally the exports would not use a single object, but multiple ES6 exports:

```js
const Foo = {…}

class Bar {…}

export { Foo, Bar }
```

…but then RN code that _does_ try to capture all exports in a single object would have to be re-written like the following, which makes this process harder to do as a static conversion (that has no context about other modules):

```ts
import * as SomeModule from "./some-module"
const { Foo } = SomeModule
```

This becomes even more problematic to convert when [the export object has lazy getters](https://github.com/facebook/react-native/blob/145151ef0f44156bcc03fa4a44a5aa630cc3e5f2/index.js#L109-L486):

```js
module.exports = {
    get Foo() {
        return require("./foo")
    }
}
```

</details>

I’m unsure yet how to proceed; mostly I think I need to sleep on it for a bit while I switch gears and get this out of my head in front of you all. Options that currently are running through my head are:

- Test a `print-signature` tool that the Flow team mentioned, which is perhaps already able to do some of this work.
- Ask the Flow team to make the CLI tool able to emit Flow declarations that use ES6 imports/exports.
- Make RN use ES6 imports/exports so these don’t need to be converted; but I’m sure there are reasons why this isn’t done yet.

The latter might be preferable anyways, as it would remove the need for `tsc` to infer untyped return types–which in turn means we can do a straight conversion and not have `tsc` introduce inline types or other such behaviour that’s hard to control.

To be continued…

**Comment by TheSavior** [comment]
> Make RN use ES6 imports/exports so these don’t need to be converted; but I’m sure there are reasons why this isn’t done yet.

Nope, no reason other than it hasn't been done. The internal Facebook codebase has been converted other than the react-native repo folder. I'm sure that landing an external PR that does this migration on the entire folder would be difficult, but if there are certain files that are causing you problems where converting those would be helpful, feel free to send PRs for those.

**Comment by alloy** [comment]
> > Make RN use ES6 imports/exports so these don’t need to be converted; but I’m sure there are reasons why this isn’t done yet.
> 
> Nope, no reason other than it hasn't been done. The internal Facebook codebase has been converted other than the react-native repo folder. I'm sure that landing an external PR that does this migration on the entire folder would be difficult, but if there are certain files that are causing you problems where converting those would be helpful, feel free to send PRs for those.

Oh, interesting! Yeah I guess I could start by doing those that problematic to convert and leave the rest to be converted automatically for now.

> Also, thank you for writing up so much of your current progress and status. I'm very excited about this so it is helpful for me to have a sense of where things are and how we can help you get unstuck. :)

You got it! 🙇
